[HUDI-9405] Adding new writer apis and impl to Metadata writer to support streaming writes

lokeshj1703 opened the pull request:
### Change Logs

The PR adds support for streaming writes in HoodieBackedMetadataTableWriter and introduces new APIs for the same. 

3 APIs are added as detailed below.

a. to start a new commit in MDT 

```
/**
   * Starts a new commit in metadata table for optimized write flow.
   * @param instantTime
   */
  void startCommit(String instantTime); 
```
b. to do streaming write to MDT 

```
/**
   * Prepare records and write to Metadata table for all eligible partitions except FILES partition. This will be used in optimized writes,
   * where in data table writes statuses are maintained as HoodieData and based on that, we prepare records and write to Metadata table
   * partitions (except FILES). Caution should be followed to ensure the action is not triggered on the incoming HoodieData < WriteStatus >
   *   and for the writes to metadata table. Caller is expected to trigger collect just once for both set of HoodieData < WriteStatus >.
   * @param writeStatus {@link HoodieData} of {@link WriteStatus} from data table writes.
   * @param instantTime instant time of interest.
   * @return {@link HoodieData} of {@link WriteStatus} for writes to metadata table.
   */
  HoodieData<WriteStatus> streamWriteToMetadataPartitions(HoodieData<WriteStatus> writeStatus, String instantTime);
```

c. to write to left over partitions in MDT including FILES partition and complete the commit. 
```
 /**
   * This api will be used in streaming writes to metadata flow, where in, a write in data table is already written to all data table, partitions in Metadata table
   * with streaming writes supported
   * using {@link #streamWriteToMetadataPartitions} and the action is triggered for all writes together.
   * Post this, after finalizing the write in data table, caller invokes this method to update metadata table for other partitions where streaming writes are not suported
   * included FILES partition and finally completing the commit in metadata table.
   *
   * @param instantTime instant time of interest.
   * @param context {@link HoodieEngineContext} of interest.
   * @param metadataWriteStatsSoFar List<HoodieWriteStat> for partial/streaming writes to metadata table completed so far.
   * @param commitMetadata {@link HoodieCommitMetadata} of interest.
   */
  void completeStreamingCommit(String instantTime, HoodieEngineContext context, List<HoodieWriteStat> metadataWriteStatsSoFar, HoodieCommitMetadata commitMetadata);
```

In this patch, we are adding the implementation for the same w/o wiring end to end. But have written tests to test them directly. In future patch, we wire these end to end. 

### Impact

Adding new APIs and impl for the same with HoodieMetadataWriter to support streaming writes. 

### Risk level (write none, low medium or high below)

none. 

### Documentation Update

_Describe any necessary documentation update if there is any new feature, config, or user-facing change. If not, put "none"._

- _The config description must be updated if new configs are added or the default value of the configs are changed_
- _Any new feature or user-facing change requires updating the Hudi website. Please create a Jira ticket, attach the
  ticket number here and follow the [instruction](https://hudi.apache.org/contribute/developer-setup#website) to make
  changes to the website._

### Contributor's checklist

- [ ] Read through [contributor's guide](https://hudi.apache.org/contribute/how-to-contribute)
- [ ] Change Logs and Impact were stated clearly
- [ ] Adequate tests were added if applicable
- [ ] CI passed

nsivabalan commented on hudi-client/hudi-java-client/src/main/java/org/apache/hudi/metadata/JavaHoodieBackedTableMetadataWriter.java:
```diff
@@ -63,6 +64,11 @@ protected JavaHoodieBackedTableMetadataWriter(StorageConfiguration<?> storageCon
 
+  @Override
+  MetadataIndexGenerator getMetadataIndexGenerator() {
+    return null;
```
same here

lokeshj1703 replied:
Addressed

nsivabalan commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/metadata/SparkHoodieBackedTableMetadataWriter.java:
```diff
@@ -144,6 +157,63 @@ protected JavaRDD<HoodieRecord> convertHoodieDataToEngineSpecificData(HoodieData
+
+  @Override
+  protected HoodieData<HoodieRecord> repartitionByMDTFileSlice(HoodieData<HoodieRecord> records, int numPartitions) {
+    return HoodieJavaRDD.of(HoodieJavaRDD.getJavaRDD(records).mapToPair(new PairFunction<HoodieRecord, Pair<String, String>, HoodieRecord>() {
```
do we have tests for this.

nsivabalan commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/metadata/SparkHoodieBackedTableMetadataWriter.java:
```diff
@@ -144,6 +157,63 @@ protected JavaRDD<HoodieRecord> convertHoodieDataToEngineSpecificData(HoodieData
+  public JavaRDD<WriteStatus> streamWriteToMetadataTable(Pair<List<Pair<String, String>>, HoodieData<HoodieRecord>> mdtRecordsHoodieData, String instantTime, boolean initialCall) {
+    JavaRDD<HoodieRecord> mdtRecords = HoodieJavaRDD.getJavaRDD(mdtRecordsHoodieData.getValue());
+
+    if (initialCall) {
```
we may not need `initialcall` arg only. 
we can just rely on mdtRecordsHoodieData and if FILES is not present, its the initial call

nsivabalan commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/metadata/SparkHoodieBackedTableMetadataWriterTableVersionSix.java:
```diff
@@ -93,6 +93,11 @@ public static HoodieTableMetadataWriter create(StorageConfiguration<?> conf, Hoo
 
+  @Override
+  MetadataIndexGenerator getMetadataIndexGenerator() {
+    return null;
```
same comment as above

nsivabalan commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/metadata/SparkHoodieBackedTableMetadataWriter.java:
```diff
@@ -144,6 +157,63 @@ protected JavaRDD<HoodieRecord> convertHoodieDataToEngineSpecificData(HoodieData
+  }
+
+  @Override
+  protected HoodieData<HoodieRecord> repartitionByMDTFileSlice(HoodieData<HoodieRecord> records, int numPartitions) {
```
why do we have similar code here and in SparkMetadataIndexGenerator.
both are trying to repartition HoodieData based on file id.

nsivabalan reviewed (changes requested):
left some high level comments.

nsivabalan reviewed (commented):
you may need to rebase on top of my latest changes to these apis. we have moved these  a separate write client named SparkRDDMetadataWriteClient.

nsivabalan commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/metadata/HoodieBackedTableMetadataWriter.java:
```diff
@@ -1094,6 +1116,137 @@ public void buildMetadataPartitions(HoodieEngineContext engineContext, List<Hood
+    List<MetadataPartitionType> mdtPartitionsToTag = new ArrayList<>(enabledPartitionTypes);
+    mdtPartitionsToTag.remove(FILES);
+    HoodieData<Pair<String, HoodieRecord>> perWriteStatusRecords = writeStatus.flatMap(
+        new MetadataIndexGenerator.PerWriteStatsBasedIndexGenerator(mdtPartitionsToTag, dataWriteConfig, storageConf, instantTime));
```
are these addressed?

lokeshj1703 commented on hudi-client/hudi-flink-client/src/main/java/org/apache/hudi/metadata/FlinkHoodieBackedTableMetadataWriter.java:
```diff
@@ -211,6 +217,11 @@ protected void upsertAndCommit(BaseHoodieWriteClient<?, List<HoodieRecord>, ?, L
 
+  @Override
+  MetadataIndexGenerator getMetadataIndexGenerator() {
+    return null;
```
Addressed

lokeshj1703 commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/metadata/MetadataIndexGenerator.java:
```diff
@@ -0,0 +1,154 @@
+    }
+  }
+
+  protected HoodieData<HoodieRecord> prepareFilesPartitionRecords(HoodieEngineContext context, HoodieCommitMetadata commitMetadata, String instantTime) {
```
Removed it

lokeshj1703 commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/metadata/HoodieBackedTableMetadataWriter.java:
```diff
@@ -1298,9 +1451,7 @@ public void update(HoodieRollbackMetadata rollbackMetadata, String instantTime)
-      HoodieInstant deltaCommitInstant = metadataMetaClient.createNewInstant(HoodieInstant.State.COMPLETED,
-          HoodieTimeline.DELTA_COMMIT_ACTION, commitToRollbackInstantTime);
-      if (metadataMetaClient.getActiveTimeline().getDeltaCommitTimeline().containsInstant(deltaCommitInstant)) {
+      if (metadataMetaClient.getActiveTimeline().getDeltaCommitTimeline().containsInstant(commitToRollbackInstantTime)) {
```
I think this is not required, we are already handling the pending commit rollback case in else block.

lokeshj1703 commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/metadata/HoodieBackedTableMetadataWriter.java:
```diff
@@ -1094,6 +1116,137 @@ public void buildMetadataPartitions(HoodieEngineContext engineContext, List<Hood
+  }
+
+  protected O streamWriteToMetadataTable(Pair<List<Pair<String, String>>, HoodieData<HoodieRecord>> taggedMdtRecords, String instantTime, boolean initialCall) {
+    throw new HoodieMetadataException("Should be implemented by engines");
```
Addressed

lokeshj1703 commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/metadata/HoodieBackedTableMetadataWriter.java:
```diff
@@ -1094,6 +1116,137 @@ public void buildMetadataPartitions(HoodieEngineContext engineContext, List<Hood
+
+    // Generate HoodieRecords for MDT partitions which need per hudi partition writeStats in one spark task
+    // for eg, partition stats index
+    HoodieData<Pair<String, HoodieRecord>> perPartitionRecords = metadataIndexGenerator.get().prepareMDTRecordsGroupedByHudiPartition(writeStatus);
```
removed it

lokeshj1703 commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/metadata/HoodieBackedTableMetadataWriter.java:
```diff
@@ -388,17 +409,17 @@ private boolean initializeFromFilesystem(String dataTableInstantTime, List<Metad
       }
     }
-    Map<String, Map<String, Long>> partitionIdToAllFilesMap = partitionInfoList.stream()
+    Map<String, Map<String, Long>> partitionToFilesMap = partitionInfoList.stream()
```
Addressed

lokeshj1703 commented:
Rebased the PR on top of the latest changes related to MDT table. Also rebased https://github.com/apache/hudi/pull/13005/commits which was required by this PR. Fixed some basic functionality around the APIs based on the latest changes in the master.

lokeshj1703 commented on hudi-client/hudi-spark-client/src/test/java/org/apache/hudi/io/TestMetadataWriterCommit.java:
```diff
@@ -0,0 +1,135 @@
+    HoodieData<WriteStatus> mdtWriteStatus = mdtWriter.streamWriteToMetadataPartitions(HoodieJavaRDD.of(Collections.singletonList(writeStatus), context, 1), instantTime);
+    List<HoodieWriteStat> mdtWriteStats = mdtWriteStatus.collectAsList().stream().map(WriteStatus::getStat).collect(Collectors.toList());
+    mdtWriter.completeStreamingCommit(instantTime, context, mdtWriteStats, commitMetadata);
+    assertEquals(3, mdtMetaClient.reloadActiveTimeline().filterCompletedInstants().countInstants());
```
Addressed

nsivabalan reviewed (commented):
I see tests for AppendHandle and CreateHandle. 
can you create one for Merge Handle as well

danny0405 commented:
@lokeshj1703 Thanks for the contribution, I have reviewed and add some cosmatic changes.

lokeshj1703 commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/metadata/HoodieTableMetadataWriter.java:
```diff
@@ -36,6 +39,37 @@
 public interface HoodieTableMetadataWriter<I,O> extends Serializable, AutoCloseable {
 
+  /**
+   * Starts a new commit in metadata table for optimized write flow.
```
Yes, in subsequent commits we would be writing to MDT using thest APIs.

lokeshj1703 replied:
Yes, in subsequent PRs for MDT writer we would be using these APIs to write to metadata table

nsivabalan replied:
we might actually fold it within our abstraction we have planned to write to metadata table. rather than BaseHoodieWriteClient directly calling metadata table write client.startCommit() directly, 
just when we are looking to start the streaming write to metadata table, the abstraction (named [HoodieMetadataWriteWrapper](https://github.com/apache/hudi/pull/13402)). 

This is to unify the flows across ingestion writers and table services. For ingestion service, we can call startCommit in mdt right away when we call write client.startcommit in data table. but for table services, we can't do that. We have to wait until the table service execution starts. So, after going back and forth, I unified the logic across both (ingestion writers and table services). 

Just when we call streaming write to metadata table, we will create a new HoodieMetadataWriter and we will start a new commit with MDT

lokeshj1703 commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/metadata/HoodieBackedTableMetadataWriter.java:
```diff
@@ -1096,6 +1119,126 @@ public void buildMetadataPartitions(HoodieEngineContext engineContext, List<Hood
+        List<FileSlice> fileSlices =
+            HoodieTableMetadataUtil.getPartitionLatestFileSlices(metadataMetaClient, Option.ofNullable(fsView), partitionName);
+        if (fileSlices.isEmpty()) {
+          // scheduling of INDEX only initializes the file group and not add commit
```
It seems like this is required for async metadata indexing.

lokeshj1703 replied:
This seems to be used for async metadata indexing. @nsivabalan do you have context on the index scheduling flow?

nsivabalan replied:
yes, when a partition is going through async indexing, ingestion writers are expected to still write log files to these partitions. for these cases, latest committed file slice might be empty and so we are polling for inflight file slices as well. 
So, generally not applicable. only incase of async indexing is inflight.

the-other-tim-brown replied:
Should this condition be updated to match that case then?

lokeshj1703 replied:
Added a condition check.

lokeshj1703 commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/metadata/SparkHoodieBackedTableMetadataWriter.java:
```diff
@@ -161,6 +185,13 @@ protected void upsertAndCommit(BaseHoodieWriteClient<?, JavaRDD<HoodieRecord>, ?
+  @Override
+  protected void upsertAndCommit(BaseHoodieWriteClient<?, JavaRDD<HoodieRecord>, ?, JavaRDD<WriteStatus>> writeClient, String instantTime, JavaRDD<HoodieRecord> preppedRecordInputs,
+                                 List<HoodieFileGroupId> mdtFileGroupsIdsToUpdate) {
+    JavaRDD<WriteStatus> writeStatusJavaRDD = ((SparkRDDMetadataWriteClient)writeClient).upsertPreppedRecords(preppedRecordInputs, instantTime, Option.of(mdtFileGroupsIdsToUpdate));
```
Addressed

lokeshj1703 commented on hudi-client/hudi-spark-client/src/test/java/org/apache/hudi/io/TestCreateHandle.java:
```diff
@@ -0,0 +1,85 @@
+/**
+ * Unit tests {@link HoodieCreateHandle}.
+ */
+@SuppressWarnings("checkstyle:LineLength")
```
Not needed. Fixed it.

nsivabalan commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/metadata/HoodieTableMetadataWriter.java:
```diff
@@ -36,6 +39,37 @@
+   * @param instantTime instant time of interest.
+   * @return {@link HoodieData} of {@link WriteStatus} for writes to metadata table.
+   */
+  HoodieData<WriteStatus> streamWriteToMetadataPartitions(HoodieData<WriteStatus> writeStatus, String instantTime);
```
Agreed. but I can't think of better naming. 
May be optimized writes is the other name I can think of. 
generally not a fan of V1, V2 naming.

danny0405 replied:
I would suggest "partialWriteToMetadataPartitions"

lokeshj1703 replied:
How about sth like `inlineWriteToMetadataPartitions`?

the-other-tim-brown commented:
Just one pending question but otherwise LGTM https://github.com/apache/hudi/pull/13286#discussion_r2140708093
